deviceplugin/usb.go

```diff
@@ -69,54 +73,85 @@ type usbDevice struct {
+	if !dir.IsDir() {
+		// There shouldn't be any raw files in this directory, but just in case.
+		return
+	}
+	// Try to find the vendor ID file inside this device - this is a good indication that we're dealing with a device, not a bus.
+	vnd, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirVendorIDFile)
+	if err != nil {
+		// We can't read the vendor file for some reason, it probably doesn't exist.
+		return
+	}
+	prd, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirProductIDFile)
+	if err != nil {
+		return
+	}
+
+	// The following two calls shouldn't fail.
+	busRaw, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirBusFile)
+	if err != nil {
+		return result, err
+	}
+	bus := binary.LittleEndian.Uint16(busRaw)
+	busLocRaw, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirBusDevFile)
+	if err != nil {
+		return result, err
+	}
+	busLoc := binary.LittleEndian.Uint16(busLocRaw)
+
+	res := usbDevice{
+		Vendor:    usbID(binary.LittleEndian.Uint16(vnd)),
+		Product:   usbID(binary.LittleEndian.Uint16(prd)),
+		Bus:       bus,
+		BusDevice: busLoc,
+	}
+	return &res, nil
+}
+
+// enumerateUSBDevices rapidly scans the OS system bus for attached USB devices.
+// Pure Go; does not require external linking.
 func enumerateUSBDevices() (specs []usbDevice, err error) {
 	allDevs, err := os.ReadDir(usbDevicesDir)
 	if err != nil {
 		return
 	}
-	for _, dev := range allDevs {
-		if !dev.IsDir() {
-			// There shouldn't be any raw files in this directory, but just in case.
-			continue
-		}
-		// Try to find the vendor ID file inside this device - this is a good indication that we're dealing with a device, not a bus.
-		vnd, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirVendorIDFile)
-		if err != nil {
-			// We can't read the vendor file for some reason, it probably doesn't exist.
-			continue
-		}
-		prd, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirProductIDFile)
-		if err != nil {
-			continue
-		}
 
-		// The following two calls shouldn't fail.
-		busRaw, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirBusFile)
-		if err != nil {
-			return specs, err
-		}
-		bus := binary.LittleEndian.Uint16(busRaw)
-		busLocRaw, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirBusDevFile)
-		if err != nil {
-			return specs, err
-		}
-		busLoc := binary.LittleEndian.Uint16(busLocRaw)
+	var wg sync.WaitGroup
+	for i, dev := range allDevs {
+		wg.Add(1)
 
-		spec := usbDevice{
-			Vendor:    usbID(binary.LittleEndian.Uint16(vnd)),
-			Product:   usbID(binary.LittleEndian.Uint16(prd)),
-			Bus:       bus,
-			BusDevice: busLoc,
-		}
-		specs = append(specs, spec)
+		// Copy the loop variables
+		index := i
+		device := dev
+
+		go func() {
+			defer wg.Done()
+			result, err := queryUSBDeviceCharacteristicsByDirectory(device)
+			if err != nil {
+				// do we want to handle errors here?
+				return
+			}
+			if result != nil {
+				specs[index] = *result
```

Comment:
So some of the specs might be empty, right? Do we want the outer block to clean up the `specs` slice? Alternatively we could receive all of the specs on a channel and add them to a slice once.

Reply:
Good catch - personally split between using a channel and just tidying the slice after the fact, I'm minded to go for a channel for performance

Reply:
Thoughts on 2ad9b37? I might be going down a complete garden path, my go toolchain has also started being wobbly (curse you darwin) and my train of thought is also being a bit wobbly tonight 😵‍💫

Reply:
yeah it's getting late :)) I'll resolve here and start a new thread in the new changes